store/decoupled/store.go

```diff
@@ -0,0 +1,161 @@
+package decoupled
+
+import (
+	"crypto/sha256"
+	"io"
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/iavl"
+	"github.com/cosmos/cosmos-sdk/store/tracekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+	// "github.com/cosmos/cosmos-sdk/telemetry"
+
+	// abci "github.com/tendermint/tendermint/abci/types"
+	dbm "github.com/tendermint/tm-db"
+)
+
+var (
+	_ types.KVStore       = (*Store)(nil)
+	_ types.CommitStore   = (*Store)(nil)
+	_ types.CommitKVStore = (*Store)(nil)
+	// _ types.Queryable               = (*Store)(nil)
+	// _ types.StoreWithInitialVersion = (*Store)(nil)
+)
+
+var (
+	versionsPrefix = []byte{0}
+	dataPrefix     = []byte{1}
+	indexPrefix    = []byte{2}
+)
+
+// A store which uses separate data structures for state storage (SS) and state commitments (SC)
+type Store struct {
+	// State commitments layer
+	sc types.CommitKVStore
+	// Direct KV mapping (SS)
+	data dbm.DB
+	// Inverted index of SC values to SS keys
+	inv dbm.DB
```

Comment:
Does the second bucket need to move through a separate tm-db interface?

Reply:
I'm using `PrefixDB`s to separate the two into different namespaces. Collisions are unlikely since the inverted index uses a hash as key, but I think it's still desired.